src/main/java/org/logstash/filewatch/WatchedFilesCollection.java

```diff
@@ -0,0 +1,229 @@
+                sortBy = "modified_at";
+                comparator = (file1, file2) -> {
+                    if (file1 == file2) return 0; // fast shortcut
+                    RubyFloat mtime1 = modified_at(context, file1);
+                    RubyFloat mtime2 = modified_at(context, file2);
+                    int cmp = Double.compare(mtime1.getDoubleValue(), mtime2.getDoubleValue());
+                    // if mtime same (rare unless file1 == file2) - order consistently
+                    if (cmp == 0) return path(context, file1).op_cmp(path(context, file2));
+                    return cmp;
+                };
+                break;
+            case "path" :
+                sortBy = "path";
+                comparator = (file1, file2) -> path(context, file1).op_cmp(path(context, file2));
+                break;
+            default :
+                throw context.runtime.newArgumentError("sort_by: '" + sort_by + "' not supported");
+        }
+
+        if ("desc".equals(sort_direction)) {
+            comparator = comparator.reversed();
+        }
+
+        this.files = new TreeMap<>(comparator);
+        this.filesInverse = RubyHash.newHash(context.runtime);
+
+        // variableTableStore("@files", JavaUtil.convertJavaToRuby(context.runtime, this.files));
+        // variableTableStore("@files_inverse", this.filesInverse);
+
+        return this;
+    }
+
+    @JRubyMethod
+    public IRubyObject add(ThreadContext context, IRubyObject file) {
+        RubyString path = getFilePath(context, file);
+        synchronized (this) {
+            RubyString prev_path = this.files.put(file, path);
+            assert prev_path == null || path.equals(prev_path); // file's path should not change!
+            this.filesInverse.op_aset(context, path, file);
+        }
+        return path;
+    }
+
+    private static RubyString getFilePath(ThreadContext context, IRubyObject file) {
+        IRubyObject path = file.callMethod(context, "path");
+        if (!(path instanceof RubyString)) {
+            throw context.runtime.newTypeError("expected file.path to return String but did not file: " + file.inspect());
+        }
+        if (!path.isFrozen()) path = ((RubyString) path).dupFrozen(); // path = path.dup.freeze
+        return (RubyString) path;
+    }
+
+    @JRubyMethod
+    public IRubyObject remove_paths(ThreadContext context, IRubyObject arg) {
+        IRubyObject[] paths;
+        if (arg instanceof RubyArray) {
+            paths = ((RubyArray) arg).toJavaArray();
+        } else {
+            paths = new IRubyObject[] { arg };
+        }
+
+        int removedCount = 0;
+        synchronized (this) {
+            for (final IRubyObject path : paths) {
+                if (removePath(context, path.convertToString())) removedCount++;
+            }
+        }
+        return context.runtime.newFixnum(removedCount);
+    }
+
+    private boolean removePath(ThreadContext context, RubyString path) {
```

Comment:
should this method be synchronized?

Since `this.filesInverse` is not a threadsafe data structure, two independent threads could succeed at deleting a file but then have alternate outcomes removing from `this.files`.

Reply:
the private piece is only used from `remove_paths` impl : 
```java
        synchronized (this) {
            for (final IRubyObject path : paths) {
                if (removePath(context, path.convertToString())) removedCount++;
            }
        }
```
have tried to do all synchronization locking at the top level (public) methods.
the private `removePath` is a "left-over" as I was re-using it at some point.
still the `remove_paths` code reads cleaner having this one around ...